### PR TITLE
DSO2250 changes.

### PR DESCRIPTION
**Pull request by phmarek**

### CODE REVIEW

**Comment by karlp** [comment]
One comment is that the segfault and leak fix should be prefixed with hantek-dso: as they're not part of the new code, and are shared bugfixes.

File: src/hardware/hantek-dso/api.c [hunk 1]
```diff
@@ -539,6 +539,7 @@ static int config_set(uint32_t key, GVariant *data,
 			devc->samplerate = samplerates[idx];
 			if (dso_set_trigger_samplerate(sdi) != SR_OK)
 				return SR_ERR;
+			sr_err("got new sample rate %d, idx %d", devc->samplerate, idx);
```
Comment (karlp) [inline src/hardware/hantek-dso/api.c]: This doesn't really look like it should be at sr_err level?

File: src/hardware/hantek-dso/protocol.c [hunk 2]
```diff
@@ -241,7 +241,187 @@ static int get_channel_offsets(const struct sr_dev_inst *sdi)
...
+{
+	struct dev_context *devc;
+	struct sr_usb_dev_inst *usb;
+	int ret, tmp;
+	int base;
+	uint8_t cmdstring[12];
+	int trig;
+
+
+	sr_dbg("Preparing CMD_2250_SET_TRIGGERSOURCE.");
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: seems not very useful given the next one at line 282?

File: src/hardware/hantek-dso/protocol.c [hunk 3]
```diff
@@ -241,7 +241,187 @@ static int get_channel_offsets(const struct sr_dev_inst *sdi)
...
+	cmdstring[0] = CMD_2250_SET_TRIGGERSOURCE;
+	sr_dbg("Trigger source %s.", devc->triggersource);
+	if (!strcmp("CH2", devc->triggersource))
+		tmp = 3;
+	else if (!strcmp("CH1", devc->triggersource))
+		tmp = 2;
+	else if (!strcmp("EXT", devc->triggersource))
+		tmp = 0;
+	else {
+		sr_err("Invalid trigger source: '%s'.", devc->triggersource);
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: this doesn't log at a higher level when it gets the SR_ERR_ARG?

File: src/hardware/hantek-dso/protocol.c [hunk 4]
```diff
@@ -241,7 +241,187 @@ static int get_channel_offsets(const struct sr_dev_inst *sdi)
...
+	if (send_begin(sdi) != SR_OK)
+		return SR_ERR;
+
+	if ((ret = libusb_bulk_transfer(usb->devhdl, DSO_EP_OUT,
+			cmdstring, 8, &tmp, 100)) != 0) {
+		sr_err("Failed to set trigger/samplerate: %s.",
+		       libusb_error_name(ret));
+		return SR_ERR;
+	}
+	sr_dbg("Sent CMD_2250_SET_TRIGGERSOURCE.");
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: this, and the one at line 325 all seem a bit spammy

File: src/hardware/hantek-dso/protocol.c [hunk 5]
```diff
@@ -241,7 +241,187 @@ static int get_channel_offsets(const struct sr_dev_inst *sdi)
...
+	memset(cmdstring, 0, sizeof(cmdstring));
+	cmdstring[0] = CMD_2250_SET_SAMPLERATE;
+	sr_dbg("Sample rate: %u", devc->samplerate);
+	base = 100e6;
+	if (devc->samplerate > base) {
+		/* Timebase fast */
+		sr_err("Sample rate > 100MHz not yet supported.");
+		return SR_ERR_ARG;
+
+		if (devc->framesize != FRAMESIZE_SMALL) {
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: So, all of this is dead code, with no indication of whether it works, or is meant to work, or is just garbage.  Just delete it.

File: src/hardware/hantek-dso/protocol.c [hunk 6]
```diff
@@ -241,7 +241,187 @@ static int get_channel_offsets(const struct sr_dev_inst *sdi)
...
+	}
+	sr_dbg("Sent CMD_2250_SET_CHANNELS.");
+
+
+
+	/* Trigger slope: 0=positive 1=negative */
+	memset(cmdstring, 0, sizeof(cmdstring));
+	cmdstring[0] = CMD_2250_SET_TRIGGERPOS_AND_BUFFER;
+
+	/* Horizontal trigger position */
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: old comment from old expectation of naming?

File: src/hardware/hantek-dso/protocol.c [hunk 7]
```diff
@@ -338,8 +521,8 @@ static int dso_set_trigger_samplerate(const struct sr_dev_inst *sdi)
 	cmdstring[5] = (tmp >> 8) & 0xff;
 
 	/* Horizontal trigger position */
-	sr_dbg("Trigger position: %3.2f.", devc->triggerposition);
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: Same old comment above

File: src/hardware/hantek-dso/protocol.c [hunk 8]
```diff
@@ -411,6 +595,80 @@ static int dso_set_voltage(const struct sr_dev_inst *sdi)
...
+	case VDIV_200MV:
+	case VDIV_20MV:
+		cmdstring[2] |= 0x04;
+		break;
+	case VDIV_5V:
+	case VDIV_500MV:
+	case VDIV_50MV:
+		cmdstring[2] |= 0x08;
+		break;
+	}
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: What's all this for? it's completely disconnected?

File: src/hardware/hantek-dso/protocol.c [hunk 9]
```diff
@@ -411,6 +595,80 @@ static int dso_set_voltage(const struct sr_dev_inst *sdi)
...
+static int dso_set_voltage(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_usb_dev_inst *usb;
+	int ret, tmp;
+	uint8_t cmdstring[8];
+
+	devc = sdi->priv;
+	if (devc->profile->fw_pid == 0x2250)
+		return dso2250_set_voltage(sdi);
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.c]: I've gotta say, I really feel like these should be setup as function pointers in the init stage or something, not sprinkled into every function, but maybe this is best for now.

File: src/hardware/hantek-dso/protocol.h [hunk 10]
```diff
@@ -131,6 +140,7 @@ enum capturestates {
 	CAPTURE_FILLING = 1,
 	CAPTURE_READY_8BIT = 2,
 	CAPTURE_READY_9BIT = 7,
+	CAPTURE_READY2250 = 3,
```
Comment (karlp) [inline src/hardware/hantek-dso/protocol.h]: are these meant to bit bit masks or what?  Either way, it seems that it should be placed  betwen 8 and 9 bit

File: src/hardware/hantek-dso/protocol.c (hunk 3 above)
Comment (phmarek) [reply src/hardware/hantek-dso/protocol.c]: No (not yet). Most of that is still left over from the non-2250 code that I copied...

**Pull request closed without merging**